lib/omniauth/strategies/facebook.rb

```diff
@@ -76,7 +76,7 @@ def callback_phase
       #      phase and it must match during the access_token phase:
       #      https://github.com/facebook/facebook-php-sdk/blob/master/src/base_facebook.php#L477
       def callback_url
-        if @authorization_code_from_signed_request_in_cookie
+        if defined?(@authorization_code_from_signed_request_in_cookie) && @authorization_code_from_signed_request_in_cookie
```

Comment:
What about to define `authorization_code_from_signed_request_in_cookie` as private `attr_accessor`?

```diff
diff --git a/lib/omniauth/strategies/facebook.rb b/lib/omniauth/strategies/facebook.rb
index 3e93f07..9b2497e 100644
--- a/lib/omniauth/strategies/facebook.rb
+++ b/lib/omniauth/strategies/facebook.rb
@@ -8,6 +8,7 @@ module OmniAuth
   module Strategies
     class Facebook < OmniAuth::Strategies::OAuth2
       class NoAuthorizationCodeError < StandardError; end
+      attr_accessor :authorization_code_from_signed_request_in_cookie
 
       DEFAULT_SCOPE = 'email'
 
@@ -76,7 +77,7 @@ module OmniAuth
       #      phase and it must match during the access_token phase:
       #      https://github.com/facebook/facebook-php-sdk/blob/master/src/base_facebook.php#L477
       def callback_url
-        if defined?(@authorization_code_from_signed_request_in_cookie) && @authorization_code_from_signed_request_in_cookie
+        if self.authorization_code_from_signed_request_in_cookie
           ''
         else
           # Fixes regression in omniauth-oauth2 v1.4.0 by https://github.com/intridea/omniauth-oauth2/commit/85fdbe117c2a4400d001a6368cc359d88f40abc7
@@ -131,7 +132,7 @@ module OmniAuth
           yield
         elsif code_from_signed_request = signed_request_from_cookie && signed_request_from_cookie['code']
           request.params['code'] = code_from_signed_request
-          @authorization_code_from_signed_request_in_cookie = true
+          self.authorization_code_from_signed_request_in_cookie = true
           # NOTE The code from the signed fbsr_XXX cookie is set by the FB JS SDK will confirm that the identity of the
           #      user contained in the signed request matches the user loading the app.
           original_provider_ignores_state = options.provider_ignores_state
@@ -140,7 +141,7 @@ module OmniAuth
             yield
           ensure
             request.params.delete('code')
-            @authorization_code_from_signed_request_in_cookie = false
+            self.authorization_code_from_signed_request_in_cookie = false
             options.provider_ignores_state = original_provider_ignores_state
           end
         else
```

Anyway I have spotted we're missing specs for this part. Do you have any idea how to test this? I would like to get this covered before any change we will do.